src/components/search/data/reducer.js

```diff
@@ -0,0 +1,70 @@
+/* eslint-disable import/prefer-default-export */
+import {
+  DELETE_REFINEMENT, SET_REFINEMENT, CLEAR_REFINEMENTS, SET_REFINEMENTS_FROM_QUERY_PARAMS, ADD_TO_REFIENEMENT_ARRAY,
+  REMOVE_FROM_REFINEMENT_ARRAY,
+} from './actions';
+import { QUERY_PARAMS_TO_IGNORE } from './constants';
+
+export const valueArrayReducer = (state = [], action) => {
+  switch (action.type) {
+    case REMOVE_FROM_REFINEMENT_ARRAY:
+      return state.filter((value) => value !== action.value);
+    case ADD_TO_REFIENEMENT_ARRAY:
+      return [...state, action.value];
+    default:
+      return [...state];
+  }
+};
+
+export const refinementsReducer = (state = {}, action) => {
+  console.info('Refinement Action Triggered \n', action);
+  const nextState = { ...state };
+  switch (action.type) {
+    case DELETE_REFINEMENT:
+      delete nextState.page;
+      delete nextState[action.key];
+      return nextState;
+    case SET_REFINEMENT:
+      delete nextState.page;
+      nextState[action.key] = action.value;
+      return nextState;
+    case SET_REFINEMENTS_FROM_QUERY_PARAMS:
+      // we don't delete the page when setting multiple refinements. This action should only be used by SearchData,
+      //  and it has to handle the page coming in as a query param. Deleting it can cause an infinite loop.
+      Object.keys(state).forEach((key) => {
+        // remove refinements that are not from query params
+        if (!QUERY_PARAMS_TO_IGNORE.includes(key)) {
+          delete nextState[key];
+        }
+      });
+      return {
+        ...nextState,
+        ...action.payload,
+      };
+    case ADD_TO_REFIENEMENT_ARRAY:
+      delete nextState.page;
+      return {
+        ...nextState,
+        [action.key]: valueArrayReducer(state[action.key], action),
```

Comment:
Is the reducer composition necessary here? What do we get out of the reducer composition over something like the following?

```jsx
case ADD_TO_REFINEMENT_ARRAY:
  delete nextState.page;
  return {
    ...nextState,
    [action.key]: [...nextState[action.key], action.value],
  };
case REMOVE_FROM_REFINEMENT_ARRAY:
  if (nextState[action.key]) {
    delete nextState.page;
    return {
      ...nextState,
      [action.key]: nextState.filter(value => value !== action.value),
    };
  }
  return nextState;
```

Reply:
My understanding is that reducer composition is best practices for making reducers. It also allows me to independently test what's happening. I kind of wanted to separate what's happening to the overall object from what's happening to the array. But you're right, it might be overkill here.

Reply:
Yeah, I get reducer composition for more complex operations, but in this case, all we're doing is concatenating and filtering an array. As for independently testing what's happening, as long as the final result is correct, that should be enough.